lib/mix/lib/mix/dep.ex

```diff
@@ -78,71 +78,115 @@ defmodule Mix.Dep do
   @doc """
   Returns loaded dependencies from the cache for the current environment.
 
+  If dependencies have not been cached yet, they are loaded
+  and then cached.
+
   Because the dependencies are cached during deps.loadpaths,
   their status may be outdated (for example, `:compile` did not
   yet become `:ok`). Therefore it is recommended to not rely
   on their status, also given they haven't been checked
   against the lock.
-
-  If MIX_NO_DEPS is set, we return an empty list of dependencies
-  without loading them.
   """
   def cached() do
-    cond do
-      System.get_env("MIX_NO_DEPS") in ~w(1 true) ->
-        []
+    if project = Mix.Project.get() do
+      read_cached_deps(project, Mix.env()) || load_and_cache()
+    else
+      load_and_cache()
+    end
+  end
+
+  @doc """
+  Returns loaded dependencies recursively and caches it.
+
+  The result is cached for future `cached/0` calls.
+
+  ## Exceptions
+
+  This function raises an exception if any of the dependencies
+  provided in the project are in the wrong format.
+  """
+  def load_and_cache() do
```

Comment:
I’m not sure about the oficial policy regarding specs for private modules, but what do you think about including `@spec`s here (and the rest of public functions included in this PR) to improve the documentation and type checking (dialyzer)?

Reply:
We only use dialyzer for documentation purposes. That's why we usually don't spec private modules.